### PR TITLE
FP8 Support for MCore MoE

### PR DESCRIPTION
**Pull request by Victarry**
Add FP8 support for MoE in MCore.

Related MR in MCore:
https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/merge_requests/1089

Implementation details:
- Add `rng_tracker_name` for initialize with EP
- Add if statement in kernel to handle zeros tokens passed to an expert.

### CODE REVIEW

**Comment by ptrendx** [comment]
I don't like the fact that the layers need to know that they are experts. Can't it be abstracted in some way using the options that we already have or add options that are more generic?

**Comment by Victarry** [comment]
I see. Good advice.

I remove the `is_expert` flag and use `explicit_parallel_comm` to indicate the communications are handled outside the te.Linear.

**Comment by Victarry** [comment]
Hi @ptrendx, could you please continue the review and share your comments. 

To make sure this feature can be included in MCore v0.6, I think it's better to merge this MR this week.

**Comment by ptrendx** [comment]
So, to be honest, I don't quite understand why we need that communication flag at all. MCore should be able to just call te.Linear without setting row or column parallelism to the same effect, no? And then we would not need any special flag on the TE side?

Also, you added this rng tracker name option, but did not document it.

Handling of the zero token case I think is fine.

**Comment by Victarry** [comment]
Added documentation. Thanks.

**Review by ptrendx — APPROVED** [review]
Changes LGTM. @Victarry, please add a test that runs the Linear layer with empty input to shw that it works and then we will be able to merge.

**Comment by Victarry** [comment]
Hi @ptrendx, I have added the new unittest for Linear layer with empty input.

Since I didn't find an appropriate file to place the new testcase, I created a new test file named `test_linear_layer.py`.
Please tell me if you have any other suggentions.

**Comment by ptrendx** [comment]
I would put it in test_sanity.py. It looks good, but please also add some check in it - like checking that the batch size of the output of the linear layer is the same as the input (so if 0 gets passed as input, 0 is also provided as the output).

**Comment by Victarry** [comment]
Done! Thanks for you advice. 👍🏻

**Comment by Victarry** [comment]
Hi, @ptrendx, can this MR be merged now?

**Comment by Victarry** [comment]
Hi, @ptrendx, I just fixed the UT error in CI, could you please trigger the ci again?

**Comment by Victarry** [comment]
Hi, @ptrendx, the CI pipeline is passed, could you please merge this MR?
Thanks a lot!

**Comment by ptrendx** [comment]
Hi @Victarry, we are trying to minimize the changes going into 1.6 release so will merge that PR after 1.6 branch is created.

**Comment by ptrendx** [comment]
Hi @Victarry, now that the 1.6 branch is created, could you resolve conflicts in your PR? Then we will be able to merge it.

**Comment by viclzhu** [comment]
Hi @Victarry, I'm wondering when the mcore related changes will be available on the public [mcore](https://github.com/NVIDIA/Megatron-LM) repository. Or if it's already available, could you point me to the relevant changes or PR? 
It sounds like it will likely be available on mcore-0.7, but I couldn't seem to find the changes yet on the [beta-0.7](https://github.com/NVIDIA/Megatron-LM/tree/core_r0.7.0.beta) branch.

Thanks!

**Comment by Victarry** [comment]
Hi, @viclzhu, the mcore related change is planed to be published before the end of May.
Thanks.

**Comment by Victarry** [comment]
Hi @ptrendx, I found that the UT only failed on L40, but I'm not sure why does this happen.

Do you have any insights?

**Comment by ptrendx** [comment]
Hi Victarry - I checked and this failure is unrelated to this PR, so I believe it is safe to merge.

File: transformer_engine/pytorch/csrc/extensions/transpose.cu [hunk 1]
```diff
@@ -335,6 +338,8 @@ at::Tensor fp8_transpose(at::Tensor input,
 
   size_t M = static_cast<size_t>(input.size(0));
   size_t N = static_cast<size_t>(input.size(1));
+  if (M == 0 || N == 0)
+    return input;
```
Comment (Victarry) [inline transformer_engine/pytorch/csrc/extensions/transpose.cu:342]: @Victarry This will cause shape mismatch error between wgrad and weight when gradient accumulation fusion is disabled.
